In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_columns',None)

In [3]:
defenses = pd.read_csv("../processed_data/weeks_frame1/frame1_process/defense_players_plays.csv",index_col=0)
defenses.head()

id      x      y     s     a   dis       o     dir      nflId  \
1  2018090600:75:1  88.89  36.47  0.01  0.01  0.01  105.63   66.66    79848.0   
3  2018090600:75:1  86.31  22.01  0.09  0.42  0.01   70.12  168.91  2495613.0   
5  2018090600:75:1  73.64  28.70  0.03  0.02  0.01  103.05  219.41  2534832.0   
7  2018090600:75:1  86.48  31.12  0.09  0.03  0.04   95.90   33.36  2552315.0   
8  2018090600:75:1  82.67  20.53  0.03  0.07  0.01   81.14  174.57  2552689.0   

       displayName  jerseyNumber position  team playDirection  
1  Malcolm Jenkins          27.0       SS  home          left  
3     Corey Graham          24.0       FS  home          left  
5    Rodney McLeod          23.0       FS  home          left  
7     Jordan Hicks          58.0      MLB  home          left  
8     Ronald Darby          21.0       CB  home          left

In [13]:
offenses = pd.read_csv("../processed_data/weeks_frame1/frame1_process/offense_players_plays.csv",index_col=0)
offenses.head()

In [4]:
plays = pd.read_csv("../processed_data/clean/plays.csv",index_col=0)
plays.head()

id                                    playDescription  \
0   2018090600:75:1  (15:00) M.Ryan pass short right to J.Jones pus...   
1  2018090600:146:1  (13:10) M.Ryan pass incomplete short right to ...   
2  2018090600:168:1  (13:05) (Shotgun) M.Ryan pass incomplete short...   
3  2018090600:190:1  (13:01) (Shotgun) M.Ryan pass deep left to J.J...   
4  2018090600:256:1  (10:59) (Shotgun) M.Ryan pass incomplete short...   

   quarter  down  yardsToGo possessionTeam        playType yardlineSide  \
0        1     1         15            ATL  play_type_pass          ATL   
1        1     1         10            ATL  play_type_pass          PHI   
2        1     2         10            ATL  play_type_pass          PHI   
3        1     3         10            ATL  play_type_pass          PHI   
4        1     3          1            ATL  play_type_pass          PHI   

   yardlineNumber offenseFormation        personnelO  defendersInTheBox  \
0              20           I_FORM  2 RB, 1 TE, 2 WR                7.0   
1              39       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
2              39          SHOTGUN  2 RB, 1 TE, 2 WR                6.0   
3              39          SHOTGUN  1 RB, 1 TE, 3 WR                6.0   
4               1          SHOTGUN  2 RB, 3 TE, 0 WR                8.0   

   numberOfPassRushers        personnelD  preSnapVisitorScore  \
0                  4.0  4 DL, 2 LB, 5 DB                  0.0   
1                  4.0  4 DL, 2 LB, 5 DB                  0.0   
2                  4.0  4 DL, 2 LB, 5 DB                  0.0   
3                  5.0  4 DL, 1 LB, 6 DB                  0.0   
4                  6.0  6 DL, 3 LB, 2 DB                  0.0   

   preSnapHomeScore gameClock  absoluteYardlineNumber penaltyCodes  \
0               0.0  15:00:00                    90.0          NaN   
1               0.0  13:10:00                    49.0          NaN   
2               0.0  13:05:00                    49.0          NaN   
3               0.0  13:01:00                    49.0          NaN   
4               0.0  10:59:00                    11.0          NaN   

  penaltyJerseyNumbers passResult  offensePlayResult  playResult       epa  \
0                  NaN          C                 10          10  0.261827   
1                  NaN          I                  0           0 -0.372360   
2                  NaN          I                  0           0 -0.702779   
3                  NaN          C                 33          33  3.047530   
4                  NaN          I                  0           0 -0.842272   

   isDefensivePI  
0          False  
1          False  
2          False  
3          False  
4          False

La zona profunda empieza después de 15 yardas de la línea de scrimmage. Estaría interesante empezar con defensas que no estén en la red zone (las últimas 20 yardas hasta la línea de anotación).

La red zone es cuando el equipo atacante está dentro de las últimas 20 yardas de su oponente.

In [5]:
plays["redZone"] = ""

In [6]:
plays.head()

id                                    playDescription  \
0   2018090600:75:1  (15:00) M.Ryan pass short right to J.Jones pus...   
1  2018090600:146:1  (13:10) M.Ryan pass incomplete short right to ...   
2  2018090600:168:1  (13:05) (Shotgun) M.Ryan pass incomplete short...   
3  2018090600:190:1  (13:01) (Shotgun) M.Ryan pass deep left to J.J...   
4  2018090600:256:1  (10:59) (Shotgun) M.Ryan pass incomplete short...   

   quarter  down  yardsToGo possessionTeam        playType yardlineSide  \
0        1     1         15            ATL  play_type_pass          ATL   
1        1     1         10            ATL  play_type_pass          PHI   
2        1     2         10            ATL  play_type_pass          PHI   
3        1     3         10            ATL  play_type_pass          PHI   
4        1     3          1            ATL  play_type_pass          PHI   

   yardlineNumber offenseFormation        personnelO  defendersInTheBox  \
0              20           I_FORM  2 RB, 1 TE, 2 WR                7.0   
1              39       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
2              39          SHOTGUN  2 RB, 1 TE, 2 WR                6.0   
3              39          SHOTGUN  1 RB, 1 TE, 3 WR                6.0   
4               1          SHOTGUN  2 RB, 3 TE, 0 WR                8.0   

   numberOfPassRushers        personnelD  preSnapVisitorScore  \
0                  4.0  4 DL, 2 LB, 5 DB                  0.0   
1                  4.0  4 DL, 2 LB, 5 DB                  0.0   
2                  4.0  4 DL, 2 LB, 5 DB                  0.0   
3                  5.0  4 DL, 1 LB, 6 DB                  0.0   
4                  6.0  6 DL, 3 LB, 2 DB                  0.0   

   preSnapHomeScore gameClock  absoluteYardlineNumber penaltyCodes  \
0               0.0  15:00:00                    90.0          NaN   
1               0.0  13:10:00                    49.0          NaN   
2               0.0  13:05:00                    49.0          NaN   
3               0.0  13:01:00                    49.0          NaN   
4               0.0  10:59:00                    11.0          NaN   

  penaltyJerseyNumbers passResult  offensePlayResult  playResult       epa  \
0                  NaN          C                 10          10  0.261827   
1                  NaN          I                  0           0 -0.372360   
2                  NaN          I                  0           0 -0.702779   
3                  NaN          C                 33          33  3.047530   
4                  NaN          I                  0           0 -0.842272   

   isDefensivePI redZone  
0          False          
1          False          
2          False          
3          False          
4          False

In [7]:
for index,row in plays.iterrows():
    playDirection = defenses[defenses["id"]==row["id"]]["playDirection"].values[0]

    
    if (row["absoluteYardlineNumber"] > 10 and row["absoluteYardlineNumber"] <=30 and playDirection == "left") or (row["absoluteYardlineNumber"] >= 90 and row["absoluteYardlineNumber"] < 110 and playDirection == "right"):
        plays.loc[plays['id']==row["id"], ["redZone"]] = True
    else:
        plays.loc[plays['id']==row["id"], ["redZone"]] = False

In [8]:
plays[(plays["redZone"]==True)]

id                                    playDescription  \
4        2018090600:256:1  (10:59) (Shotgun) M.Ryan pass incomplete short...   
9        2018090600:521:1  (6:19) M.Ryan pass short left to A.Hooper to P...   
10       2018090600:545:1          (5:38) M.Ryan pass incomplete short left.   
11       2018090600:636:1  (5:07) (Shotgun) M.Ryan pass incomplete short ...   
21      2018090600:1202:1  (9:07) (No Huddle, Shotgun) N.Foles pass short...   
...                   ...                                                ...   
19045  2018123014:3870:17  (7:02) N.Mullens pass short left to K.Bourne f...   
19055   2018123015:418:17  (8:15) J.Rosen pass incomplete deep middle to ...   
19056   2018123015:462:17  (7:30) (Shotgun) J.Rosen pass incomplete deep ...   
19074  2018123015:1585:17  (5:48) J.Rosen pass short right to L.Fitzgeral...   
19110  2018123015:3734:17  (6:59) (Shotgun) R.Wilson sacked at ARZ 24 for...   

       quarter  down  yardsToGo possessionTeam        playType yardlineSide  \
4            1     3          1            ATL  play_type_pass          PHI   
9            1     1         10            ATL  play_type_pass          PHI   
10           1     2          7            ATL  play_type_pass          PHI   
11           1     3          3            ATL  play_type_pass          PHI   
21           2     2         10            PHI  play_type_pass          ATL   
...        ...   ...        ...            ...             ...          ...   
19045        4     3          2             SF  play_type_pass           LA   
19055        1     1         10            ARI  play_type_pass          SEA   
19056        1     3         11            ARI  play_type_pass          SEA   
19074        2     1         10            ARI  play_type_pass          SEA   
19110        4     3          9            SEA  play_type_sack          ARI   

       yardlineNumber offenseFormation        personnelO  defendersInTheBox  \
4                   1          SHOTGUN  2 RB, 3 TE, 0 WR                8.0   
9                  10       SINGLEBACK  1 RB, 2 TE, 2 WR                8.0   
10                  7       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
11                  3          SHOTGUN  1 RB, 2 TE, 2 WR                6.0   
21                 20          SHOTGUN  1 RB, 1 TE, 3 WR                7.0   
...               ...              ...               ...                ...   
19045               2       SINGLEBACK  1 RB, 1 TE, 3 WR                8.0   
19055              16       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
19056              17          SHOTGUN  1 RB, 1 TE, 3 WR                6.0   
19074              15       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
19110              20          SHOTGUN  1 RB, 1 TE, 3 WR                7.0   

       numberOfPassRushers        personnelD  preSnapVisitorScore  \
4                      6.0  6 DL, 3 LB, 2 DB                  0.0   
9                      3.0  4 DL, 2 LB, 5 DB                  0.0   
10                     4.0  4 DL, 2 LB, 5 DB                  0.0   
11                     4.0  4 DL, 2 LB, 5 DB                  0.0   
21                     4.0  3 DL, 3 LB, 5 DB                  3.0   
...                    ...               ...                  ...   
19045                  5.0  4 DL, 2 LB, 5 DB                 17.0   
19055                  5.0  4 DL, 2 LB, 5 DB                  0.0   
19056                  6.0  4 DL, 1 LB, 6 DB                  0.0   
19074                  4.0  4 DL, 2 LB, 5 DB                  3.0   
19110                  5.0  4 DL, 2 LB, 5 DB                 21.0   

       preSnapHomeScore gameClock  absoluteYardlineNumber penaltyCodes  \
4                   0.0  10:59:00                    11.0          NaN   
9                   0.0  06:19:00                    20.0          NaN   
10                  0.0  05:38:00                    17.0          NaN   
11                  0.0  05:07:00      

In [9]:
len(plays[(plays["redZone"]==True)])

2534

Hay un total de 2534 jugadas en la redzone

In [10]:
len(plays[(plays["absoluteYardlineNumber"]<=30)&(plays["possessionTeam"]!=plays["yardlineSide"])&(plays["redZone"]==True)]) + len(plays[(plays["absoluteYardlineNumber"]>=90)&(plays["possessionTeam"]!=plays["yardlineSide"])])

2534

In [14]:
plays[plays["redZone"]==True]

id                                    playDescription  \
4        2018090600:256:1  (10:59) (Shotgun) M.Ryan pass incomplete short...   
9        2018090600:521:1  (6:19) M.Ryan pass short left to A.Hooper to P...   
10       2018090600:545:1          (5:38) M.Ryan pass incomplete short left.   
11       2018090600:636:1  (5:07) (Shotgun) M.Ryan pass incomplete short ...   
21      2018090600:1202:1  (9:07) (No Huddle, Shotgun) N.Foles pass short...   
...                   ...                                                ...   
19045  2018123014:3870:17  (7:02) N.Mullens pass short left to K.Bourne f...   
19055   2018123015:418:17  (8:15) J.Rosen pass incomplete deep middle to ...   
19056   2018123015:462:17  (7:30) (Shotgun) J.Rosen pass incomplete deep ...   
19074  2018123015:1585:17  (5:48) J.Rosen pass short right to L.Fitzgeral...   
19110  2018123015:3734:17  (6:59) (Shotgun) R.Wilson sacked at ARZ 24 for...   

       quarter  down  yardsToGo possessionTeam        playType yardlineSide  \
4            1     3          1            ATL  play_type_pass          PHI   
9            1     1         10            ATL  play_type_pass          PHI   
10           1     2          7            ATL  play_type_pass          PHI   
11           1     3          3            ATL  play_type_pass          PHI   
21           2     2         10            PHI  play_type_pass          ATL   
...        ...   ...        ...            ...             ...          ...   
19045        4     3          2             SF  play_type_pass           LA   
19055        1     1         10            ARI  play_type_pass          SEA   
19056        1     3         11            ARI  play_type_pass          SEA   
19074        2     1         10            ARI  play_type_pass          SEA   
19110        4     3          9            SEA  play_type_sack          ARI   

       yardlineNumber offenseFormation        personnelO  defendersInTheBox  \
4                   1          SHOTGUN  2 RB, 3 TE, 0 WR                8.0   
9                  10       SINGLEBACK  1 RB, 2 TE, 2 WR                8.0   
10                  7       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
11                  3          SHOTGUN  1 RB, 2 TE, 2 WR                6.0   
21                 20          SHOTGUN  1 RB, 1 TE, 3 WR                7.0   
...               ...              ...               ...                ...   
19045               2       SINGLEBACK  1 RB, 1 TE, 3 WR                8.0   
19055              16       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
19056              17          SHOTGUN  1 RB, 1 TE, 3 WR                6.0   
19074              15       SINGLEBACK  1 RB, 1 TE, 3 WR                7.0   
19110              20          SHOTGUN  1 RB, 1 TE, 3 WR                7.0   

       numberOfPassRushers        personnelD  preSnapVisitorScore  \
4                      6.0  6 DL, 3 LB, 2 DB                  0.0   
9                      3.0  4 DL, 2 LB, 5 DB                  0.0   
10                     4.0  4 DL, 2 LB, 5 DB                  0.0   
11                     4.0  4 DL, 2 LB, 5 DB                  0.0   
21                     4.0  3 DL, 3 LB, 5 DB                  3.0   
...                    ...               ...                  ...   
19045                  5.0  4 DL, 2 LB, 5 DB                 17.0   
19055                  5.0  4 DL, 2 LB, 5 DB                  0.0   
19056                  6.0  4 DL, 1 LB, 6 DB                  0.0   
19074                  4.0  4 DL, 2 LB, 5 DB                  3.0   
19110                  5.0  4 DL, 2 LB, 5 DB                 21.0   

       preSnapHomeScore gameClock  absoluteYardlineNumber penaltyCodes  \
4                   0.0  10:59:00                    11.0          NaN   
9                   0.0  06:19:00                    20.0          NaN   
10                  0.0  05:38:00                    17.0          NaN   
11                  0.0  05:07:00      

In [18]:
plays[plays["redZone"]==True]["id"].values

array(['2018090600:256:1', '2018090600:521:1', '2018090600:545:1', ...,
       '2018123015:462:17', '2018123015:1585:17', '2018123015:3734:17'],
      dtype=object)

In [21]:
len(defenses[defenses["id"].isin(plays[plays["redZone"]==True]["id"].values)])

19234

In [26]:
len(defenses[defenses["id"].isin(plays[plays["redZone"]==False]["id"].values)])

127270

In [27]:
len(defenses) == len(defenses[defenses["id"].isin(plays[plays["redZone"]==False]["id"].values)]) + len(defenses[defenses["id"].isin(plays[plays["redZone"]==True]["id"].values)])

True

In [24]:
len(offenses[offenses["id"].isin(plays[plays["redZone"]==True]["id"].values)])

15140

In [ ]:
len(offenses[offenses["id"].isin(plays[plays["redZone"]==False]["id"].values)])

In [28]:
len(offenses) == len(offenses[offenses["id"].isin(plays[plays["redZone"]==True]["id"].values)]) + len(offenses[offenses["id"].isin(plays[plays["redZone"]==False]["id"].values)])

True

In [29]:
defenses_rz = defenses[defenses["id"].isin(plays[plays["redZone"]==True]["id"].values)]
defenses_no_rz = defenses[defenses["id"].isin(plays[plays["redZone"]==False]["id"].values)]

In [30]:
offenses_rz = offenses[offenses["id"].isin(plays[plays["redZone"]==True]["id"].values)]
offenses_no_rz = offenses[offenses["id"].isin(plays[plays["redZone"]==False]["id"].values)]

In [32]:
defenses_rz.to_csv("../processed_data/processing_defenses/defense_players_plays_rz.csv")
defenses_no_rz.to_csv("../processed_data/processing_defenses/defense_players_plays_no_rz.csv")

In [31]:
offenses_rz.to_csv("../processed_data/processing_defenses/offense_players_plays_rz.csv")
offenses_no_rz.to_csv("../processed_data/processing_defenses/offense_players_plays_no_rz.csv")

In [34]:
plays.to_csv("../processed_data/processing_defenses/plays.csv")